src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -23662,7 +23662,138 @@ public void Main()
         }
 
         [Fact]
-        public void AssertsTrue_NotNull()
+        public void DoesNotReturn_Simple()
+        {
+            string lib_cs = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [DoesNotReturn] public static void Throw() => throw null!;
+}
+";
+            string source = @"
+class D
+{
+    void Main(object? o)
+    {
+        C.Throw();
+        o.ToString(); // unreachable so no warning
+    }
+}
+";
+            var lib = CreateNullableCompilation(new[] { lib_cs, DoesNotReturnAttributeDefinition });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics();
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, DoesNotReturnAttributeDefinition });
+            comp2.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void DoesNotReturn_Operator()
+        {
+            // Annotations not honored on user-defined operators yet https://github.com/dotnet/roslyn/issues/32671
+            string lib_cs = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [DoesNotReturn] public static C operator +(C a, C b) => throw null!;
+}
+";
+            string source = @"
+class D
+{
+    void Main(object? o, C c)
+    {
+        _ = c + c;
+        o.ToString(); // unreachable so no warning
+    }
+}
+";
+            var lib = CreateNullableCompilation(new[] { lib_cs, DoesNotReturnAttributeDefinition });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(
+                // (7,9): warning CS8602: Dereference of a possibly null reference.
+                //         o.ToString(); // unreachable so no warning
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "o").WithLocation(7, 9)
+                );
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, DoesNotReturnAttributeDefinition });
+            comp2.VerifyDiagnostics(
+                // (7,9): warning CS8602: Dereference of a possibly null reference.
+                //         o.ToString(); // unreachable so no warning
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "o").WithLocation(7, 9)
+                );
+        }
+
+        [Fact]
+        public void DoesNotReturn_WithDoesNotReturnIf()
+        {
+            // warn on misuse of attribute (on a method that returns a value)? https://github.com/dotnet/roslyn/issues/36073
```

Comment:
>warn on misuse of attribute (on a method that returns a value)? [](start = 15, length = 63)

Why generate a warning? What if this is an override?

Reply:
Good point. This could be a valid scenario.

---
In reply to: [297990688](https://github.com/dotnet/roslyn/pull/36810#discussion_r297990688) [](ancestors = 297990688)